File: Go-DSA-Ejemplos/algorithms/recursion/fibonacci.go
package recursion

import (
	"fmt"
	"time"
)

// --- 1. Enfoque Recursivo (Ingenuo) ---
func FibRecursive(n int) int {
	if n <= 1 {
		return n
	}
	return FibRecursive(n-1) + FibRecursive(n-2)
}

// --- 2. Enfoque Iterativo (Eficiente) ---
func FibIterative(n int) int {
	if n <= 1 {
		return n
	}
	a, b := 0, 1
	for i := 2; i <= n; i++ {
		next := a + b
		a = b
		b = next
	}
	return b
}

// --- 3. Enfoque Recursivo con Memoización (Optimizado) ---
var memo = make(map[int]int)

func FibMemoized(n int) int {
	if n <= 1 {
		return n
	}
	if val, ok := memo[n]; ok {
		return val
	}

	result := FibMemoized(n-1) + FibMemoized(n-2)
	memo[n] = result
	return result
}

func main() {
	num := 35 // Para números grandes, la recursiva ingenua será muy lenta

	fmt.Printf("Calculando el Fibonacci de %d con diferentes métodos:\n\n", num)

	// Iterativo (el más rápido)
	startTime := time.Now()
	resultIter := FibIterative(num)
	endTime := time.Now()
	fmt.Printf("Iterativo:     %d (Tiempo: %.6f segundos)\n", resultIter, endTime.Sub(startTime).Seconds())

	// Memoizado
	memo = make(map[int]int) // Limpiar memoización para cada ejecución
	startTime = time.Now()
	resultMemo := FibMemoized(num)
	endTime = time.Now()
	fmt.Printf("Memoizado:     %d (Tiempo: %.6f segundos)\n", resultMemo, endTime.Sub(startTime).Seconds())

	// Recursivo (el más lento, puede tardar mucho)
	fmt.Println("\nEl cálculo recursivo puro puede ser muy lento, ten paciencia...")
	startTime = time.Now()
	resultRec := FibRecursive(num)
	endTime = time.Now()
	fmt.Printf("Recursivo puro: %d (Tiempo: %.6f segundos)\n", resultRec, endTime.Sub(startTime).Seconds())
}

File: Rust-DSA-Ejemplos/src/algorithms/recursion/fibonacci.rs
//! Módulo para demostrar el cálculo de la secuencia de Fibonacci en Rust.

use std::collections::HashMap;
use std::time::Instant;

// --- 1. Enfoque Recursivo (Ingenuo) ---
// Calcula el n-ésimo número de Fibonacci usando recursión simple.
pub fn fib_recursive(n: u64) -> u64 {
    if n <= 1 {
        return n;
    }
    fib_recursive(n - 1) + fib_recursive(n - 2)
}

// --- 2. Enfoque Iterativo (Eficiente) ---
// Calcula el n-ésimo número de Fibonacci usando un bucle.
pub fn fib_iterative(n: u64) -> u64 {
    if n <= 1 {
        return n;
    }
    let mut a = 0;
    let mut b = 1;
    for _ in 2..=n {
        let next = a + b;
        a = b;
        b = next;
    }
    b
}

// --- 3. Enfoque Recursivo con Memoización (Optimizado) ---
// Calcula el n-ésimo número de Fibonacci usando recursión con memoización.
pub fn fib_memoized(n: u64, memo: &mut HashMap<u64, u64>) -> u64 {
    if n <= 1 {
        return n;
    }
    if let Some(&val) = memo.get(&n) {
        return val;
    }

    let result = fib_memoized(n - 1, memo) + fib_memoized(n - 2, memo);
    memo.insert(n, result);
    result
}

// Función de demostración para Fibonacci
pub fn fibonacci_demo() {
    let num = 35; // Para números grandes, la recursiva ingenua será muy lenta

    println!("Calculando el Fibonacci de {} con diferentes métodos:\n", num);

    // Iterativo (el más rápido)
    let start_time = Instant::now();
    let result_iter = fib_iterative(num);
    let end_time = Instant::now();
    println!("Iterativo:     {} (Tiempo: {:.6?} segundos)", result_iter, end_time.duration_since(start_time));

    // Memoizado
    let mut memo = HashMap::new(); // Limpiar memoización para cada ejecución
    let start_time = Instant::now();
    let result_memo = fib_memoized(num, &mut memo);
    let end_time = Instant::now();
    println!("Memoizado:     {} (Tiempo: {:.6?} segundos)", result_memo, end_time.duration_since(start_time));

    // Recursivo (el más lento, puede tardar mucho)
    println!("\nEl cálculo recursivo puro puede ser muy lento, ten paciencia...");
    let start_time = Instant::now();
    let result_rec = fib_recursive(num);
    let end_time = Instant::now();
    println!("Recursivo puro: {} (Tiempo: {:.6?} segundos)", result_rec, end_time.duration_since(start_time));
}
